libgadget/density.c

```diff
@@ -374,17 +405,24 @@ density_ngbiter(
                   " We haven't implemented tracer particles and this shall not happen\n");
     }
 
+    /* some performance measures*/
+    O->Ninteractions ++;
+
     if(r2 < iter->kernel.HH)
     {
-
         const double u = r * iter->kernel.Hinv;
         const double wk = density_kernel_wk(&iter->kernel, u);
+        O->Ngb += wk * iter->kernel_volume;
+
+        /* For the BH only Ngb is used*/
+        if(I->Type == 5)
```

Comment:
?? BH model uses density. Did you recompute it later?

Reply:
Ah. Haha. It does! But because this was not in Reduce, this was reading from an uninitialised variable all along! Doh! Nice catch!

Reply:
No it looks like blackhole.c recomputes the density based on the Hsml computed in density.c -- thus we do not need density of BH here at all. Actually we could have splited density.c to two tree walks, Hsml that works on BH and Gas, and  sphdensity that only works on Gas, and caleed after Hsml iteration is done.

Reply:
Oh no, that's not true. BH Density is computed in blackhole.c, during accretion treewalk. The Density computed here is not used. The current code is fine. I have added a comment.

Reply:
I did try (hard) to split the density.c into two tree walks, one for BH and one for gas but it was unfortunately always slower.

Reply:
That's surprising -- it seems like we only do less work by splitting - we can avoid the redundant calculation of rho, dhsml, grad rho etc in the iterations. 

BTW, there is an almost identical DM Hsml solver in winds.c, (for the vdisp dependent winds).

Reply:
That's surprising -- it seems like we only do less work by splitting - we can avoid the redundant calculation of rho, dhsml, grad rho etc in the iterations. 

BTW, there is an almost identical DM Hsml solver in winds.c, (for the vdisp dependent winds).

Reply:
Yes, it surprised me too! It didn't seem to be tree walk overhead, so I think it had something to do with the caches.

Reply:
I'll make a note to try unifying the DM and Gas HSML code at some point - we can't run winds in the same loop as density() because we need to know which are the new stars. But maybe we can reuse some routines. Or maybe move BH Hsml into blackhole accretion.